File: c/armstrong-numbers/armstrong_numbers.c
#include "armstrong_numbers.h"
#include <math.h>

const int N = 7;

bool is_armstrong_number(int candidate) {
    if (candidate < 10) { return true; }

    int num = candidate;
    int sum = 0;
    int d = 0;
    int digits[N];

    while (num) {
        digits[d] = num % 10;
        num /= 10;
        d++;
    }

    for (int i = 0; i < d; i++) {
        sum += pow(digits[i], d);
    }

    return sum == candidate;
}

File: fortran/armstrong-numbers/armstrong_numbers.f90

module armstrong_numbers
  implicit none
contains

  ! solution without intermediate array
  logical function isArmstrongNumber(n)
    integer, intent(in) :: n
    integer :: nd, m, s

    ! single digit numbers are armstrong numbers
    if (n < 10) then
      isArmstrongNumber = .true.
      return
    end if

    nd = floor(log10(real(n))) + 1 ! number of digits
    m = n  ! since n should not be writeable
    s = 0  ! current sum of digits

    ! add up powers of digits
    do while (m > 0)
      s = s + mod(m, 10) ** nd
      m = m/10
    end do

    isArmstrongNumber = n == s

  end function

end module
